expression/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+use std::{iter, ops};
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+pub mod conversion;
+pub mod display;
+pub mod visitors;
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub enum AlgebraicExpression<T, R> {
+    Reference(R),
+    Number(T),
+    BinaryOperation(AlgebraicBinaryOperation<T, R>),
+    UnaryOperation(AlgebraicUnaryOperation<T, R>),
+}
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub struct AlgebraicBinaryOperation<T, R> {
+    pub left: Box<AlgebraicExpression<T, R>>,
+    pub op: AlgebraicBinaryOperator,
+    pub right: Box<AlgebraicExpression<T, R>>,
+}
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub enum AlgebraicBinaryOperator {
+    Add,
+    Sub,
+    Mul,
+}
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub struct AlgebraicUnaryOperation<T, R> {
+    pub op: AlgebraicUnaryOperator,
+    pub expr: Box<AlgebraicExpression<T, R>>,
+}
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub enum AlgebraicUnaryOperator {
+    Minus,
+}
+
+impl<T, R> AlgebraicExpression<T, R> {
+    /// Returns an iterator over all (top-level) expressions in this expression.
+    /// This specifically does not implement Children because otherwise it would
```

Comment:
```suggestion
    /// This specifically does not implement the Children trait because otherwise it would
```
(not sure if it is obvious in this crate)